generators/typescript/codegen/src/ast/TypeLiteral.ts

```diff
@@ -0,0 +1,255 @@
+import { assertNever } from "@fern-api/core-utils";
+import { AstNode, Writer } from "./core";
+import { Type } from "./Type";
+
+type InternalTypeLiteral = Array_ | Boolean_ | Number_ | Object_ | String_ | Tuple;
+
+interface Array_ {
+    type: "array";
+    valueType: Type;
+    fields: ArrayField[];
+    multiline: boolean;
+    leftBrace: "[";
+    rightBrace: "]";
+}
+
+interface Boolean_ {
+    type: "boolean";
+    value: boolean;
+}
+
+interface Number_ {
+    type: "number";
+    value: number;
+}
+
+interface Object_ {
+    type: "object";
+    fields: ObjectField[];
+    multiline: boolean;
+    leftBrace: "{";
+    rightBrace: "}";
+}
+
+interface String_ {
+    type: "string";
+    value: string;
+}
+
+interface Tuple {
+    type: "tuple";
+    fields: TupleField[];
+    multiline: boolean;
+    leftBrace: "[";
+    rightBrace: "]";
+}
+
+interface IterableLiteral<T extends IterableLiteralField> {
+    fields: T[];
+    multiline?: boolean;
+    leftBrace: string;
+    rightBrace: string;
+}
+
+type IterableLiteralField = ArrayField | ObjectField | TupleField;
+
+interface ArrayField {
+    type: "arrayField";
+    value: TypeLiteral;
+}
+
+interface ObjectField {
+    type: "objectField";
+    name: string;
+    valueType: Type;
+    value: TypeLiteral;
+}
+
+interface TupleField {
+    type: "tupleField";
+    valueType: Type;
+    value: TypeLiteral;
+}
+
+export class TypeLiteral extends AstNode {
+    private constructor(public readonly internalType: InternalTypeLiteral) {
+        super();
+    }
+
+    public write(writer: Writer): void {
+        switch (this.internalType.type) {
+            case "array": {
+                this.writeIterable(writer, this.internalType);
+                break;
+            }
+            case "boolean": {
+                writer.write(this.internalType.value.toString());
+                break;
+            }
+            case "number": {
+                // N.B. Defaults to decimal; further work needed to support alternatives like hex, binary, octal, etc.
+                writer.write(this.internalType.value.toString());
+                break;
+            }
+            case "object": {
+                this.writeIterable(writer, this.internalType);
+                break;
+            }
+            case "string": {
+                if (this.internalType.value.includes("\n")) {
+                    writer.write("`");
+                    const parts = this.internalType.value.split("\n");
+                    const head = parts[0] + "\n";
+                    const tail = parts.slice(1).join("\n");
+                    writer.write(head.replaceAll("`", "\\`"));
+                    writer.writeNoIndent(tail.replaceAll("`", "\\`"));
+                    writer.write("`");
+                } else {
+                    writer.write(`"${this.internalType.value.replaceAll('"', '\\"')}"`);
+                }
+                break;
```

Comment:
Nice - for readability, what if we made a couple helper functions for this? I imagine something like:

```typescript
if (this.internalType.value.includes("\n") {
  this.writeStringWithBackticks(this.internalType.value);
  break;
}
this.writeStringWithQuotes(this.internalType.value);
break;
```

Then I imagine the helper might be as simple as the following:

```typescript
function writeStringWithBackticks({writer, value} : {writer: Writer; value: string; }) {
  writer.write("`");
  writer.writeNoIndent(this.internalType.value.replaceAll("`", "\\`"));
  writer.write("`");
}
```

Similarly, `writeStringWithQuotes` could use the same structure:

```typescript
function writeStringWithBackticks({writer, value} : {writer: Writer; value: string; }) {
  writer.write('"');
  writer.write(this.internalType.value.replaceAll('"', '\\"'));
  writer.write('"');
}
```

Reply:
Discussed offline - this actually handles multi-line strings better than I expected and we should do something similar elsewhere. Nicely done!

Reply:
done